### PR TITLE
add assignment of agents to boxes

### PR DESCRIPTION
**Pull request by Westh**

### CODE REVIEW

File: parser/parser.go [hunk 1]
```diff
@@ -165,9 +165,55 @@ func preproccessLvl(levelInfo *level.Info, state *level.CurrentState) {
...
+				}
+
+				coordsToConsider := agent.Coordinates
+
+				boxesAssignedToAgent, isAnotherBoxAssignedToAgent := agentBoxAssignment[agent.Letter]
+				if isAnotherBoxAssignedToAgent {
+					coordsToConsider = boxGoalAssignment[boxesAssignedToAgent[len(boxesAssignedToAgent)-1]]
+				}
+
+				cost := level.ManhattanPlusPlus(coordsToConsider, box.Coordinates, state, &box)
```
Comment (alexdor) [inline parser/parser.go:201]: Considering that function has the `isAgentAndBoxTogetherLikeBros` part, should we maybe create another function that doesn't include it and call that function here?

File: parser/parser.go [hunk 2]
```diff
@@ -165,9 +165,55 @@ func preproccessLvl(levelInfo *level.Info, state *level.CurrentState) {
 
 	wg.Wait()
 
+	agentBoxAssignment := make(map[byte][]int)
+
+	func() {
```
[the remark this answers is not in the document]
Comment (alexdor) [reply parser/parser.go:170]: Also can this part run in the goroutine that is calculating boxGoalAssignment, once that calculation is completed?

File: parser/parser.go [hunk 3]
```diff
@@ -165,9 +165,55 @@ func preproccessLvl(levelInfo *level.Info, state *level.CurrentState) {
...
+		isSingleAgent := len(state.Agents) == 1
+		if isSingleAgent {
+			agentLetter := state.Agents[0].Letter
+
+			for boxIndex, _ := range state.Boxes {
+				agentBoxAssignment[agentLetter] = append(agentBoxAssignment[agentLetter], boxIndex)
```
[the remark this answers is not in the document]
Comment (alexdor) [reply parser/parser.go:176]: In theory, we could just skip it completely for single agent. Keep the single agent as is and just do assignments for multi

**Comment by Westh** [comment]
> Looks good but before we merge it I think we need to cherry-pick the following (in order to resolve the issue that was caused from the revert)
> 
> [44a9d8d](https://github.com/alexdor/dtu-ai-mas-final-assignment/commit/44a9d8d20b11f91580905a27e0423000dc6d578d)
> 
> [a56d130](https://github.com/alexdor/dtu-ai-mas-final-assignment/commit/a56d130012158ebf90a242a959ccf299884110a1)
> 
> [d3f8e6d](https://github.com/alexdor/dtu-ai-mas-final-assignment/commit/d3f8e6debf1d4d7642fe6638455cb69749eb7783)
> 
> what do you think?

What about just merging the whole PR?

File: level/current.go [hunk 4]
```diff
@@ -112,9 +112,19 @@ outer:
 }
 
 func (c *CurrentState) IsBoxAndCanMove(coor Coordinates, agentChar byte) bool {
-	for _, box := range c.Boxes {
-		if box.Coordinates == coor {
-			return c.LevelInfo.AgentColor[agentChar] == c.LevelInfo.BoxColor[box.Letter]
+	isSingleAgent := len(c.Agents) > 1
```
[the remark this answers is not in the document]
Comment (Westh) [reply level/current.go:115]: haha lol

File: level/cost.go [hunk 5]
```diff
@@ -29,7 +33,15 @@ func CalculateAggregatedCost(currentState *CurrentState) int {
 		}
 		goal := goals[goalIndex]
 
-		aggregatedCost += ManhattanPlusPlus(box.Coordinates, goal, currentState, &box)
+		manHattanCost := ManhattanPlusPlus(box.Coordinates, goal, currentState, &box)
+		aggregatedCost += manHattanCost
```
Comment (alexdor) [inline level/cost.go:37]: small diff but shouldn't this be after the if statement?

**Pull request merged**